### ISSUE TITLE
tidy-requires: module paths w/in require subforms should sort 

### ISSUE THREAD

**Issue by Bogdanp** [jackfirth/resyntax#432]
[This suggestion](https://github.com/Bogdanp/racket-sentry/pull/1/files#diff-ad73206235863982e1c159370b73068636233ac25168f41b116f9f043af662b6R3-R11) conflicts with [this racket-review rule](https://github.com/Bogdanp/racket-review/blob/5e0ed770fc7c649e1d5164a4815136c526780fa7/lint.rkt#L797-L835) which ensures that uses of require subforms sort according to their module path. Another example can be found [here](https://github.com/Bogdanp/racket-http-easy/pull/30/files#diff-d2e2ed902fe12e272acb437a29942ed4cbc60337aac1082e2011f2339d0ff61eL3-L6).

**Comment by jackfirth** [comment]
I agree with this, and others have mentioned the possibility to me too. But I'd also like to keep Resyntax and `racket-mode` in sync when it comes to `require` tidying. I'll reach out to Greg to discuss this.

**Comment by greghendershott** [comment]
This sounds good in principle. I'd need to reload my brain with details about `multi-in`; the other clauses seem straightforward. I'll comment back here later...

**Comment by greghendershott** [comment]
So actually. Having finished my first coffee. I realized that `racket-tidy-requires` already works the way you described in your email, as in your example:

```racket
(require a
         b
         c
         (only-in a/extra x))
```

becomes

```racket
(require a
         (only-in a/extra x)
         b
         c)
```

Was there maybe some other example where it does not? Maybe I have a bug in some cases. If so, I'll fix. But the intent is definitely to sort by module (within each meta level).

TL;DR: I think you could go ahead and change resyntax to sort that way, and if some bug turns up, then it's on me to fix that.

**Comment by jackfirth** [comment]
Ah, I was going off the description [here](https://www.racket-mode.com/#racket_002dtidy_002drequires) rather than actually installing `racket-mode` and seeing what it does, since I'm not an emacs user. What does it do if the `require` subform doesn't contain a single collection in it? Like `(prefix-in foo: (combine-in a b c))`, for example.

**Comment by Bogdanp** [comment]
Review didn't support `combine-in` at all (because I never use it), so I just added it:

https://github.com/Bogdanp/racket-review/commit/6ce349fabb80ec565c7cc22b74cf16bcd3fbed8c#diff-dd9e24dda16c8ecc44babd298601cf79cbe4cd068711b7e6f8b8135691492e2a

Running tidy requires on that same file results in:

```racket
#lang racket/base
(require (prefix-in combined: (combine-in threading))
         (combine-in)
         racket/string
         (prefix-in combined: (combine-in threading racket/base)))
```

**Comment by greghendershott** [comment]
> Ah, I was going off the description [here](https://www.racket-mode.com/#racket_002dtidy_002drequires) rather than actually installing `racket-mode` and seeing what it does, since I'm not an emacs user.

Although that's intended as user docs, not a full spec, it is ambiguous; I can improve that.

> What does it do if the `require` subform doesn't contain exactly one collection in it? Like `(prefix-in foo: (combine-in a b c))`, for example.

It's not even attempting to handle `combine-in`, yet.

I started a branch last week to refactor some of this (very old) code of mine, looking at the `require` grammar.

Among other things, I did add `combine-in` -- but I didn't consider the empty `(combine-in)` case. Thanks for pointing it out!

My instinct is to prune empty `(combine-in)` clauses. Although "tidy" shouldn't try to prune _unused imports_ -- that's the job of a "stronger" command in Racket Mode, that runs the unused requires analysis) -- pruning no-meaning, no-op "import nothing" clauses like `(combine-in)` is probably OK or in fact desirable. What do you think?

---

Generally: It looks like I should re-read the grammar carefully, looking for any more edge cases or bad assumptions.

**Comment by jackfirth** [comment]
Perhaps the three of us ought to standardize on a shared implementation? A minimal-dependency package that we all reuse would get rid of any inconsistency problems.

**Comment by greghendershott** [comment]
I'm also not opposed to code sharing. But ATM I'm rewriting some of the old code. So that it still passes the old tests. But also handles some holes like `combine-in`. And is less smelly, to me, in some spots.

When the dust settles, I can touch base re sharing.

Few quick points about that, meanwhile:

- My "outer API" is text => test: My Racket back end gets strings from Emacs, and returns strings. That doesn't mean a symbol or even syntax API couldn't be added, as well. But status quo, that's my edge data.

- Racket Mode attempts to support old versions of Racket. Currently as old as 6.12.

- Racket Mode is installed as an Emacs package, not a Racket package. It doesn't depend on any `raco setup`. Some optional features work if the necessary libs were installed by the user, otherwise they gracefully-ish don't work. Tidying requires fits fine in this category, I think. Already it needs `macro-debugger-lib` installed.

- For a certain private project I had an infatuation with `multi-in`, and added support for the simpler flavors of that. That complicates my current code, somewhat, for better or worse.

**Comment by greghendershott** [comment]
IMHO what `review` does is reasonably a "subset" of what resyntax or Racket Mode might do. IOW the latter might format or arrange things in a way that's nice, but doesn't necessarily rise to the level of a code review warning or recommendation. I think?

It's probably more important for `resyntax` and Racket Mode to be in sync.

Also to be clear, I don't even know how my Racket Mode users actively use `racket-tidy-requires`, `racket-trim-requires`, and `racket-base-requires`. And when a project uses `resyntax`, my feelings wouldn't be hurt if the project forbid using those Racket Mode commands on source code, and lets `resyntax` prevail, when it's being applied to all the code all the time.

**Comment by greghendershott** [comment]
I took a fresh look at the old "tidying" code and came up with this, squashed commit: https://github.com/greghendershott/racket-mode/commit/ecde6e8b51495916ea6ee873349b3988b40e5c18

It tries to be more principled about following the require spec grammar.

It also tries to reduce or elide useless forms -- even just as part of just basic tidying. (This is independent of the "stronger" business of using `macro-debugger-lib` to identify modules that needn't be imported at all, which IIUC is a non-goal for resyntax.)

IIUC the sorting shouldn't contradict what you're doing @Bogdanp in racket-review.

@jackfirth I'm not sure if this is something that you would like to reuse?

If so, I could look at making this available as a package (although I might "vendor" that same source directly into the repo for Racket Mode... I would need to think about the pros and cons).

**Comment by jackfirth** [comment]
@greghendershott For Resyntax to reuse it, it would have to use syntax objects instead of plain datums. Resyntax relies on syntax properties and source locations in the inputs and outputs of refactoring rules to figure various things out. So whatever tidying system Resyntax uses has to accept and return a `(require ...)` syntax object.

**Comment by greghendershott** [comment]
In that commit using syntax classes to drive most of the work, I do store syntax objects in the "sortable" structs.

Currently I convert to datums because the result of my internal `tidy` function going into a `tidy-pretty-format` function anyway -- currently my outer API returns a string, with everything pre-indented.

I could trivially avoid conversion back to datums. Probably a bit more work to ensure that I'm manipulating syntax objects without losing srcloc and props, _everywhere_. But probably straightforward.

Two questions:

1. How does resyntax handle indentation?  (If you get a stx obj, how/when do you convert back to text?)

2. How does resyntax handle the case where the original program has multiple `require` forms within a module?

   In my notion of "tidy", these ought to be combined into one `require` form. And so the API is you're responsible for giving me a list of zero or more old require forms that you found in the program, deleting them all, and inserting at the site of the first one the new pre-formatted text I give back to you.

   And so I wonder what syntax object(s) you would desire to supply, as well as to get back?

I guess with both questions, my meta-question is, "What's your ideal API?"  Using syntax objects, OK, but other details like this?

**Comment by greghendershott** [comment]
I think part of the impedance mismatch is that `racket-tidy-requires` doesn't just _sort_ existing `require` forms.

It also "gathers" multiple `require` forms within a module, and consolidates them into just one.

So if some file has

```
(require a)
(require b)
(require c)
```

it will produce

```
(require a b c)
```

Also, the `racket-base-requires` variant -- which adds `require`s to change from `#lang racket` to `racket/base` -- can add a `require` form when the original file had zero.

So it is pretty basic to the API that the result is some sort of "change list" -- a list of inserts and deletes. The special case of a delete+insert at the same position, can be presented instead as a replace (which can help the client handle surrounding whitespace more nicely). So the change list consists of "delete", "insert", and "replace".

---

I think what this means is that resyntax, status quo, isn't attempting to do "as much" as `racket-tidy-requires`. And maybe never wants to attempt more? I don't know.

I think this also means is that any shared library would need to offer an API that allows access to the "sorting" aspect -- but not "gathering"/consolidating.

(And of course, you'd want that to work as syntax => syntax. But that alone isn't the biggest deal.)
